### PR TITLE
Add support for chained and retry task

### PR DESCRIPTION
**Pull request by angxu**
Add two useful util methods for task composition.
- Task#andThen()
- Task#retry()

### CODE REVIEW

File: controller/controller-worker/src/main/java/com/pinterest/rocksplicator/controller/tasks/DelayAckTaskQueue.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+ * {@link #finishTask(long, String)} and {@link #failTask(long, String)}.
+ *
+ * It enables ChainedTask to achieve ack-and-enqueue operation atomically.
+ */
+class DelayAckTaskQueue implements TaskQueue {
+  private final TaskQueue taskQueue;
+  private final AtomicReference<State> stateRef;
+
+  public static class State {
+    public static int UNFINISHED = 0;
```
Comment (shuzhang1989) [inline controller/controller-worker/src/main/java/com/pinterest/rocksplicator/controller/tasks/DelayAckTaskQueue.java]: enum?

File: controller/controller-worker/src/main/java/com/pinterest/rocksplicator/controller/tasks/ChainedTask.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Stack;
+
+/**
+ * @author Ang Xu (angxu@pinterest.com)
+ */
```
Comment (angxu) [inline controller/controller-worker/src/main/java/com/pinterest/rocksplicator/controller/tasks/ChainedTask.java]: Sure.

File: controller/controller-worker/src/main/java/com/pinterest/rocksplicator/controller/tasks/DelayAckTaskQueue.java (hunk 1 above)
Comment (angxu) [inline controller/controller-worker/src/main/java/com/pinterest/rocksplicator/controller/tasks/DelayAckTaskQueue.java]: Sure.

File: controller/controller-worker/src/main/java/com/pinterest/rocksplicator/controller/tasks/TaskBase.java [hunk 3]
```diff
@@ -84,4 +84,32 @@ public Task getBean() throws JsonProcessingException {
...
+   * Returns a chained task of this task and {@code nextTask} in a way that if this task returns
+   * successfully, the nextTask will be processed immediately after this task.
+   * @param nextTask
+   * @return a chained task
+   * @throws JsonProcessingException
+   */
+  public TaskBase andThen(TaskBase nextTask) throws JsonProcessingException {
+    return new ChainedTask(
+        new ChainedTask.Param()
+            .setT1(this.getBean())
```
Comment (angxu) [inline controller/controller-worker/src/main/java/com/pinterest/rocksplicator/controller/tasks/TaskBase.java]: Not really. The following code is also possible. 
```
TaskBase t1, t2;
new ChainedTask(t1, t2);
```

**Pull request closed without merging**